plugins/woocommerce/src/Blocks/Utils/ProductCollectionUtils.php

```diff
@@ -92,4 +99,179 @@ private static function remove_empty_array_recursive( $array ) {
 		}
 		return $array;
 	}
+
+	/**
+	 * Parse editor's location context from WP Post.
+	 *
+	 * Possible contexts:
+	 * - post
+	 * - page
+	 * - single-product
+	 * - product-archive
+	 * - cart
+	 * - checkout
+	 * - product-catalog
+	 * - order-confirmation
+	 *
+	 * @param WP_Post $post The Post instance.
+	 *
+	 * @return string Returns the context.
+	 */
+	public static function parse_editor_location_context( $post ) {
+		$context = 'site';
+
+		if ( ! $post instanceof \WP_Post ) {
+			return $context;
+		}
+
+		$post_type = $post->post_type;
+		if ( ! in_array( $post_type, array( 'post', 'page', 'wp_template', 'wp_template_part' ), true ) ) {
+			return $context;
+		}
+
+		if ( in_array( $post_type, array( 'wp_template', 'wp_template_part' ), true ) ) {
+
+			$name = $post->post_name;
+			if ( false !== strpos( $name, SingleProductTemplate::SLUG ) ) {
+				$context = 'single-product';
+			} elseif ( ProductAttributeTemplate::SLUG === $name ) {
+				$context = 'product-archive';
+			} elseif ( false !== strpos( $name, 'taxonomy-' ) ) { // Including the '-' in the check to avoid false positives.
+				$taxonomy           = str_replace( 'taxonomy-', '', $name );
+				$product_taxonomies = get_object_taxonomies( 'product', 'names' );
+				if ( in_array( $taxonomy, $product_taxonomies, true ) ) {
+					$context = 'product-archive';
+				}
+			} elseif ( in_array( $name, array( CartTemplate::SLUG, MiniCartTemplate::SLUG ), true ) ) {
+				$context = 'cart';
+			} elseif ( CheckoutTemplate::SLUG === $name ) {
+				$context = 'checkout';
+			} elseif ( ProductCatalogTemplate::SLUG === $name ) {
+				$context = 'product-catalog';
+			} elseif ( OrderConfirmationTemplate::SLUG === $name ) {
+				$context = 'order-confirmation';
+			}
+		}
+
+		if ( 'page' === $post_type ) {
+			$context = 'page';
+		}
+		if ( 'post' === $post_type ) {
+			$context = 'post';
+		}
+
+		return $context;
+	}
+
+	/**
+	 * Track usage of the Product Collection block within the given blocks.
+	 *
+	 * @param array $blocks     The parsed blocks to check.
+	 * @param bool  $in_single  Whether we are in a single product container (used for keeping state in the recurring process).
+	 *
+	 * @return array Parsed instances of the Product Collection block.
+	 */
+	public static function parse_blocks_track_data( $blocks, $in_single = false ) {
+
+		$instances = array();
+
+		foreach ( $blocks as $block ) {
+			if ( 'woocommerce/product-collection' === $block['blockName'] ) {
+
+				$instances[] = array(
+					'collection'        => $block['attrs']['collection'] ?? 'product-catalog',
+					'in-single-product' => $in_single ? '1' : '0',
+					'filters'           => self::get_query_filters_track_data( $block ),
+				);
+			}
+
+			$local_in_single = $in_single;
```

Comment:
```suggestion
			$is_in_single_product = $in_single;
```

Reply:
Thanks for the suggestion here!

Naming was tricky here since this function is recurring, with `$in_single` carrying the current result and the other variable prefixed with `local_[name]` indicating it will hold the result for the next iteration. 

Perhaps a better name would be:
- `$is_in_single_product`
- `$local_is_in_single_product`

Thoughts? Would that help with readability?

Reply:
Both looks good to me. Feel free to use any of those names. It will improve the readability of the code 🙌🏻

Reply:
Done in https://github.com/woocommerce/woocommerce/pull/46466/commits/40f84d47b864f9d6323b7436159f5b58656ea20c